### ISSUE TITLE
Issue in introducing multi tenancy in sftpgo

### ISSUE THREAD

**Issue by sunilke** [drakkan/sftpgo#648]

**Comment by drakkan** [comment]
Hi,

I know some proprietary software that use something similar but also start several "worker" processes (those that serve SFTP / FTP etc ...), so user1 belonging to vsite1 must use a port other than user2 belonging to vsite2 to access the SFTP service . To have something similar you can externally launch and manage multiple SFTPGo instances. This keeps users isolated and you can set up a reverse proxy if you want to expose the admin interface for the different instances on a single external port. This approach ensures greater isolation. Your approach might also work, its advantage is that the super admin can see everything, but virtual sites aren't isolated.

**Comment by sunilke** [comment]
Sure, understood. Will evaluate this approach too. Thanks for quick reply

**Comment by drakkan** [comment]
You are welcome, the only thing that I care about is that the code, which you will eventually write, is publicly available so that I can evaluate and eventually merge it (with or without changes) into the main repository now or in the future. Thanks for your interest in SFTPGo

**Comment by sunilke** [comment]
Yeah , for sure that's the main motive.

**Comment by drakkan** [comment]
I am thinking about this feature.

**Preliminary notes:**

- protocols like SFTP/FTP have no tenant concept, so we cannot have "user1" in "tenant1" and "user1" in "tenant2", usernames must remain globally unique. So you have to use something like email addresses as usernames or add a tenant-specific suffix or prefix to uniquely identify tenant resources
- for the supported protocols several settings are configurable only at startup for example the supported SFTP ciphers, so you cannot have a tenant that supports an old cipher and another in which this old cipher is disabled. Protocol settings are global.

If the above limitations are not acceptable you have to setup different SFTPGo instances listining on different ports. This is what many proprietary software do under the hood, but I don't want to create and manage system services (this will not work well in Docker for example), so if you want something like this you have to manually setup the different instances yourself.

**Multi tenancy support**

A new "tenants" table will be added, it just contains "name" and "description".
This table will be empty by default.

We'll add the following field to the "admins" table:

- tenant_id, optional foreign key to the "tenants" table

We'll add a new "manage_tenants" permissions for the admins. This way existing admins with the "*" permission will automatically have also this permission, no migration is required. An admin cannot remove the "manage_tenants" permission to itself.

Each administrator-managed database table (users,folders,api_keys) will have an optional foreign key to the "tenants" table. API keys refer users or admins but by using a direct field we avoid a join (which must be emulated for non sql data providers).
The defender's tables will be global.
The "active_transfers" table is handled internally and will remain global.
The "shares" table does not need the tenant_id, it is managed by users, not by admins, each user can define its own shares.

You can continue using SFTPGo as you do now without adding any tenant. This way the default tenant is implicit and includes all resources with no tenant_id.

Ad admin can only manage one tenant.
We cannot allow an admin to manage multiple tenants, otherwise we have to handle situations where an admin adds a folder referenced to "tenant1" to a user referenced to "tenant2" and this is not allowed in multi tenant applications where each tenant should be isoltated from the others.

Each admin will only see the active connections for users with the same tenant_id.
Defender entries are not related to any tenant, we can have an IP banned just because it connects without trying to logging in so all the admins will be able to see all the defender entries.

When an admin create a new resource it will be automatically associated to its tenant (that is empty by default).
For existing installation all the resources will remain without a tenant id and so they belong to the default tenant. You can create a new tenant and add new resources to them.

In summary:

- an admin with no tenant id will manage resources with no tenant id
- an admin with tenant_id = 1 will manage resources with tenant_id = 1 and so on

No exception will be allowed no admin can manage multiple tenants.

These are my first thoughts on this feature. If / when I start coding it, something might be implemented a little differently and of course I could miss something.

Are you/your company interested in sponsoring this development?

**Comment by sunilke** [comment]
Hello , 
We had also finalized design almost similar to this and I had also started something but after then we decided to just use plain sftpgo without multi tenancy where we only have sftpgo users and no concept of tenant because of certain organizational reasons that came up.

Currently sftpgo isn't deployed on production for us and we are planning to deploy it sometime in future(might be in next 2-3 months) ,also for start we will go without multi tenancy whenever it will be deployed .

If we might again have requirement for multi tenancy(in case if our first prod release is successful and catering to our needs) then I will surely ask my company for sponsoring this development as it sound really interesting.
